### ISSUE TITLE
elpy goto-definition fails

### ISSUE THREAD

**Issue by sonictk** [jorgenschaefer/elpy#888]
Hi:

I've been trying to get a good setup for Emacs and Python going for a while now.

After struggling with company-jedi for a while, I decided to give elpy another go, since I heard it has better virtualenv support. Completion with company works great. However, when I try to use the goto-definition function, it fails, and I get the following error:

elpy-rpc--call-blocking: Timeout during RPC call get_definition from backend

And it fails to go to the definition of the symbol. 

Also, another question: is it possible to do the pyvenv-activate on a virtual environment and have all the features included (navigation, completion etc.) without having to install jedi/rope/flycheck/yapf/etc. into that virtual environment? It kind of defeats the purpose of a virtual environment that is controlled and isolated if it always needs to have those packages installed in order to be used with the IDE...

Would greatly appreciate any help with setting up elpy properly here! 

(This is on Windows 7, Python 2.7.6 installed. All dependencies are installed in system python and accessible.)

**Comment by jorgenschaefer** [comment]
Hello!

To the first question, that's an error in yapf – could you try to either upgrade the module, or remove it? If the latest version exhibits the error, this should be reported upstream.

To the second one:

> is it possible to do the pyvenv-activate on a virtual environment and have all the features included (navigation, completion etc.) without having to install jedi/rope/flycheck/yapf/etc. into that virtual environment?

For jedi to work correctly, you have to make both the modules you work on and the jedi module available to a single Python interpreter. Installing them all in the same virtualenv is the easiest way to do this. There are a few other options, though. You can install jedi etc. globally (or as a user) and use virtualenvwrapper's `togglesitepackages` to include globally installed packages in your virtualenv. Or you can install them in a separate virtualenv and set `PYTHONPATH` to include that virtualenv.

> It kind of defeats the purpose of a virtual environment that is controlled and isolated if it always needs to have those packages installed in order to be used with the IDE...

I'm not so sure this is true – you also install test frameworks, mock libraries etc. in your virtualenvs. But well, matter of taste. :-)

**Comment by sonictk** [comment]
@jorgenschaefer Hi, thanks for the speedy response! I did upgrade all the dependencies to their latest versions (though strangely when I use elpy-config, it showed that no yapf version was found, though every other package seemed fine. And I can import yapf from a Python shell as well). 

On a slightly related note, I wonder if it is necessary for yapf to be installed in order to do navigation to function definitions though? I am hoping to use this config on my workstation at work as well eventually but we are locked to Python 2.6 there and I can't get yapf to work on it since it is only officially supported by 2.7 and above.

**Comment by jorgenschaefer** [comment]
> I wonder if it is necessary for yapf to be installed in order to do navigation to function definitions though?

No it is not. If yapf is not installed, auto-formatting won't work (or use autopep8 if _that_ is installed), but that's it.

**Comment by sonictk** [comment]
Hi @jorgenschaefer:

I did update yapf to the latest package, and ran both --user and global site package install, but I still get the same issue, and can't goto definitions for code using elpy. If I do pip list, I see it clearly installed (0.8.2).

**Comment by jorgenschaefer** [comment]
If you run a Python shell in that virtualenv and do `from yapf.yapflib import yapf_api`, do you get an error, too?

**Comment by sonictk** [comment]
Actually I just uninstalled yapf and now it works O__O. Could you check what version of yapf you have installed on your own workstation by any chance and I'll install that specific version to see if that's the issue?

Also, as an aside, for the backend, if I set it to automatic, if both rope and jedi are available, how do I know which one it chooses to use?

**Comment by jorgenschaefer** [comment]
> Actually I just uninstalled yapf and now it works O__O.

Yep – the problem is that Elpy imports yapf, and importing yapf already throws an error. That's why I said initially that deinstalling would be a workaround. :-)

> Could you check what version of yapf you have installed on your own workstation by any chance and I'll install that specific version to see if that's the issue?

yapf (0.8.2)

> Also, as an aside, for the backend, if I set it to automatic, if both rope and jedi are available, how do I know which one it chooses to use?

It will use rope in this case.

**Comment by sonictk** [comment]
Strangely enough, even after uninstalling it...elpy-config shows it installed now. And at 0.8.2.

I'm...just going to leave it for now since it's working and I don't use yapf anywhere in my codebase. If I ever need to actually install it, I'll try to figure out what the issue is then...?

**Comment by jorgenschaefer** [comment]
You can check which virtualenv Elpy is using in `elpy-config`, too, and try `import yapf ; yapf.__file__` in that virtualenv – it's not like Elpy makes these things up :-D
